pkg/component/controller/workerconfig/reconciler.go

```diff
@@ -0,0 +1,606 @@
+/*
+Copyright 2022 k0s authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workerconfig
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"path/filepath"
+	"reflect"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/k0sproject/k0s/pkg/apis/k0s.k0sproject.io/v1beta1"
+	"github.com/k0sproject/k0s/pkg/applier"
+	"github.com/k0sproject/k0s/pkg/component"
+	"github.com/k0sproject/k0s/pkg/component/controller/leaderelector"
+	workerconfig "github.com/k0sproject/k0s/pkg/component/worker/config"
+	"github.com/k0sproject/k0s/pkg/constant"
+	kubeutil "github.com/k0sproject/k0s/pkg/kubernetes"
+
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	kubeletv1beta1 "k8s.io/kubelet/config/v1beta1"
+	"k8s.io/utils/pointer"
+
+	"github.com/sirupsen/logrus"
+	"golang.org/x/exp/slices"
+	"sigs.k8s.io/yaml"
+)
+
+type resources = []*unstructured.Unstructured
+
+type mtex = sync.Mutex
+
+// Reconciler maintains ConfigMaps that hold configuration to be
+// used on k0s worker nodes, depending on their selected worker profile.
+type Reconciler struct {
+	log logrus.FieldLogger
+
+	cleaner       cleaner
+	leaderElector leaderelector.Interface
+
+	mu            mtex
+	state         interface{}
```

Comment:
can be done instead by a custom type over string, like usually custom keys for context are implemented, for example. Easier to read + some type of safery from compiler. 

https://medium.com/@matryer/context-keys-in-go-5312346a868d

Though there is only one option for state field, why not make it more specific?

Reply:
There are actually four states: created, initialized, started and stopped. They each contain different data. Newtypes over strings will be able to express the current state's "display name", but wouldn't be able to hold the data associated with each state.

Since the states themselves don't share any commonalities, there's no natural way of letting them implement a common interface that could be used as the `state` variable's type to make it more strict. There's of course the possibility of using something along the lines of `type reconcilerState interface { isReconcilerState() }` and implement that for each of the states to mimic some sort of a "marker" interface, but not sure if that's a meaningful approach.

Btw: What's the correct name for those custom `type Foo Bar` alias types in Golang? This pattern is called "newtypes" in other programming languages. Does Go use the same name, or is there another term for those?